binnacle/lib/providers/bluetooth.dart

```diff
@@ -115,65 +130,74 @@ class BluetoothManager {
     deviceConnection = null;
     device = null;
     isConnected.add(_isConnected);
-
   }
 
-
-  // _readCharacteristic(BluetoothCharacteristic c) async {
-  //   await device.readCharacteristic(c);
-  // }
+  // t
 
   // _writeCharacteristic(BluetoothCharacteristic c) async {
   //   await device.writeCharacteristic(c, [0x12, 0x34],
   //       type: CharacteristicWriteType.withResponse);
-    
+
   // }
 
   // _readDescriptor(BluetoothDescriptor d) async {
   //   await device.readDescriptor(d);
-    
+
   // }
 
   // _writeDescriptor(BluetoothDescriptor d) async {
   //   await device.writeDescriptor(d, [0x12, 0x34]);
-    
-  // }
 
-  // _setNotification(BluetoothCharacteristic c) async {
-  //   if (c.isNotifying) {
-  //     await device.setNotifyValue(c, false);
-  //     // Cancel subscription
-  //     valueChangedSubscriptions[c.uuid]?.cancel();
-  //     valueChangedSubscriptions.remove(c.uuid);
-  //   } else {
-  //     await device.setNotifyValue(c, true);
-  //     // ignore: cancel_subscriptions
-  //     final sub = device.onValueChanged(c).listen((d) {
-  //         // print('onValueChanged $d');
-  //     });
-  //     // Add to map
-  //     valueChangedSubscriptions[c.uuid] = sub;
-  //   }
   // }
 
+  _setNotification(BluetoothCharacteristic c) async {
+    if (c.isNotifying) {
+      await device.setNotifyValue(c, false);
+      // Cancel subscription
+      await valueChangedSubscriptions[c.uuid]?.cancel();
+      valueChangedSubscriptions.remove(c.uuid);
+    } else {
+      await device.setNotifyValue(c, true);
+      // ignore: cancel_subscriptions
+      final sub = device.onValueChanged(c).listen((d) {
+        bluetoothDataStream.addStream(_test(String.fromCharCodes(d)));
+
+        bluetoothDataStream.listen((d) => print(d.toString()));
+        // print('!!!' + String.fromCharCodes(d));
+      });
+      // Add to map
+      valueChangedSubscriptions[c.uuid] = sub;
+    }
+  }
+
+  Stream _test(input) async* {
```

Comment:
Does this have to be asynchronous? I'm assuming so because it's being called from another async function, but I just wanted to make sure.

Reply:
yup